src/scipp/core/binning.py

```diff
@@ -0,0 +1,494 @@
+    """Compute a histogram, skipping NaN values.
+
+    Like :py:func:`scipp.hist`, but NaN values are skipped. See there for details and
+    examples.
+    """
+    edges = _make_edges(x, arg_dict, kwargs)
+    if len(edges) > 0:
+        x = x.bin(edges)
+    return x.bins.nansum()
+
+
+def bin(x: Union[_cpp.DataArray, _cpp.Dataset],
+        arg_dict: Dict[str, Union[int, Variable]] = None,
+        /,
+        **kwargs: Union[int, Variable]) -> Union[_cpp.DataArray, _cpp.Dataset]:
+    """Create binned data by binning input along all dimensions given by edges.
+
+    Bin edges can be specified in three ways: (1) When an integer is provided,
+    a 'linspace' with this requested number of bins is created, based on the min
+    and max of the corresponding coordinate. (2) A scalar scipp variable (a value
+    with a unit) is interpreted as a target bin with, and an 'arange' covering the
+    min and max of the corresponding coordinate is create. (3) A custom coordinate,
+    given as a scipp variable with compatible unit. Typically this should have a single
+    dimension matching the target dimension.
+
+    When binning a dimension with an existing dimension-coord, the binning for
+    the dimension is modified, i.e., the input and the output will have the same
+    dimension labels.
+
+    When binning by non-dimension-coords, the output will have new dimensions
+    given by the names of these coordinates. These new dimensions replace the
+    dimensions the input coordinates depend on.
+
+    Parameters
+    ----------
+    x:
+        Input data.
+    arg_dict:
+        Dictionary mapping dimension labels to binning parameters.
+    **kwargs:
+        Mapping of dimension label to correspoding binning parameters.
+
+    Returns
+    -------
+    :
+        Binned data.
+
+    See Also
+    --------
+    scipp.hist:
+        For histogramming data.
+    scipp.group:
+        Creating binned data by grouping, instead of binning based on edges.
+    scipp.binning.make_binned:
+        Lower level function that can bin and group, and does not automatically
+        replace/erase dimensions.
+
+    Examples
+    --------
+
+    Bin a table by one of its coord columns, specifying (1) number of bins, (2)
+    bin width, or (3) actual binning:
+
+      >>> table = sc.data.table_xyz(100)
+      >>> da = table.bin(x=2)
+      >>> da = table.bin(x=sc.scalar(0.2, unit='m'))
+      >>> da = table.bin(x=sc.linspace('x', 0.2, 0.8, num=10, unit='m'))
+
+    Bin a table by two of its coord columns:
+
+      >>> table = sc.data.table_xyz(100)
+      >>> da = table.bin(x=4, y=6)
+
+    Bin binned data, using new bins along existing dimension:
+
+      >>> binned = sc.data.binned_x(nevent=100, nbin=10)
+      >>> da = binned.bin(x=20)
+
+    Bin binned data along an additional dimension:
+
+      >>> binned = sc.data.binned_x(nevent=100, nbin=10)
+      >>> da = binned.bin(y=5)
```

Comment:
Would this be the place to add examples on the erasing behaviour or do we want to write a dedicated page on the behaviour? Maybe that would be better, and we can then just link from here to that page?

It would be nice to see the outputs of what is being done without having to check it against a reference output. I know we've talked about this before, but it would be nice to disable that check on a case by case basis.

Could we use plotting instead of string output?

Reply:
Maybe have the `sizes` as output?

Reply:
I did this in my latest push, I think it is indeed better than with no output at all.